components/compositing/compositor.rs

```diff
@@ -1465,6 +1465,23 @@ impl<Window: WindowMethods> IOCompositor<Window> {
 
             // Paint the scene.
             self.webrender.render(self.frame_size);
+
+            // Collect the currently painted epoch of each pipeline that is
+            // complete (i.e. has *all* layers painted to the requested epoch).
+            let mut pipeline_epochs = HashMap::new();
+            for (id, _) in &self.pipeline_details {
+                let webrender_pipeline_id = id.to_webrender();
+                if let Some(webrender_api::Epoch(epoch)) =
+                    self.webrender.current_epoch(webrender_pipeline_id) {
+                        let epoch = Epoch(epoch);
+                        pipeline_epochs.insert(*id, epoch);
+                    }
+            }
+
+            let msg = ConstellationMsg::PaintedEpochs(pipeline_epochs);
+            if let Err(e) = self.constellation_chan.send(msg) {
```

Comment:
An effect of this design (collect epochs, send them to constellation, then send info to layout) is that the time recorded for the metric is not when compositing finished - it is when the layout thread finished processing the notification that compositing finished.